### PR TITLE
DDPB-3259 - Upload queued documents to Sirius from daemon command

### PR DESCRIPTION
**Pull request by acsauk-mojd**
## Purpose
This ticket adds the plumbing into the existing sync document command to actually send the document to API Gateway and persist the returned UUID against he report submission and sync status against the Document.

Fixes DDPB-3259.

## Approach
There are a few callouts that happen as part of this command:

`S3` - to get the document contents
`Sirius API Gateway` - To send the Report Document and retrieve the UUID and send supporting documents once the report Document has been successfully sent.
`API` - to persist the result of the sync against the Document

I've tried to keep the logic for making these calls encapsulated in the Entity they relate to but the function is still a bit larger than I would have liked. I considered extracting some of the error handling out into their own classes but I'm not sure if it would increase the readability enough to warrant it.

## Learning
_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about DigiDeps_

## Checklist
- [x] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [x] I have added tests to prove my work
- [ ] The product team have approved these changes

### Frontend
- [ ] I have run an in-browser accessibility test (e.g. WAVE, Lighthouse)
- [ ] There are no deprecated CSS classes noted in the profiler
- [ ] Translations are used and the profiler doesn't identify any missing

### CODE REVIEW

File: client/app/config/services.yml [hunk 1]
```diff
@@ -45,10 +45,31 @@ services:
...
 
     Aws\Ssm\SsmClient:
         arguments: ["%ssm_client_params%"]
 
     AppBundle\Service\FeatureFlagService:
         arguments: ['@Aws\Ssm\SsmClient', '%env(FEATURE_FLAG_PREFIX)%']
+
+    AppBundle\Serializer\SiriusDocumentUploadSerializer:
+        autowire: true
+        autoconfigure: true
```
Comment (acsauk-mojd) [inline client/app/config/services.yml]: Still not sure why autoconfig and autowire are being ignored in defaults.yml...

File: client/src/AppBundle/Service/DocumentSyncService.php [hunk 2]
```diff
@@ -24,87 +29,223 @@ class DocumentSyncService
     /** @var S3Storage */
     private $storage;
 
-    public function __construct(S3Storage $storage)
+    /**
+     * @var SiriusApiGatewayClient
+     */
+    private $siriusApiGateWayClient;
```
Comment (gregtyler) [inline client/src/AppBundle/Service/DocumentSyncService.php]: A minor thing, but the casing of this isn't quite right
```suggestion
    private $siriusApiGatewayClient;
```

File: client/src/AppBundle/Service/DocumentSyncService.php [hunk 3]
```diff
@@ -24,87 +29,223 @@ class DocumentSyncService
...
-        } else if (in_array($report->getType(), [Report::TYPE_HEALTH_WELFARE, Report::TYPE_COMBINED_HIGH_ASSETS, Report::TYPE_COMBINED_LOW_ASSETS])) {
-            $type = 'HW';
+        if (!$document->supportingDocumentCanBeSynced()) {
+            return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_QUEUED);
+        }
+
+        $content = $this->retrieveDocumentContentFromS3($document);
+
+        if ($content instanceof Throwable) {
+            return null;
```
Comment (gregtyler) [inline client/src/AppBundle/Service/DocumentSyncService.php]: This should use `handleDocumentStatusUpdate` to set status/error message like the report PDF does. Might be worth including that setting in `retrieveDocumentContentFromS3`, `handleSiriusSync` rather than handling it in the sync functions?

Alternatively, something like this (which I prefer because I'm not a fan of returning errors from functions):
```
try {
  $content = $this->retrieveDocumentContentFromS3($document);
  $this->handleSiriusSync($document, $content);
  return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_SUCCESS);
} catch (Throwable $e) {
  return $this->handleDocumentStatusUpdate($document, Document::SYNC_STATUS_PERMANENT_ERROR, $e->getMessage());
}
```

File: client/src/AppBundle/Entity/Report/Document.php [hunk 4]
```diff
@@ -343,6 +353,26 @@ public function isAdminDocument()
...
+    {
+        if ($this->isReportPdf()) {
+            return $this->getReportSubmission();
+        }
+
+        foreach ($this->getReport()->getReportSubmissions() as $submission) {
+            if ($submission->hasReportPdf()) {
+                return $submission;
+            }
+        }
```
Comment (gregtyler) [inline client/src/AppBundle/Entity/Report/Document.php]: I'm not sure this will handle resubmissions (after OPG has returned the report to the deputy) properly. It looks like it will always return the _first_ submission with a PDF when we probably want it to return the _previous_.

So we might need to do something more like:
```suggestion
        $previousSubmissions = array_filter($this->getReport()->getReportSubmissions(), function($submission) { $submission->createdOn <= $this->getReportSubmission()->createdOn; });
        foreach (array_reverse($previousSubmissions) as $submission) {
            if ($submission->hasReportPdf()) {
                return $submission;
            }
        }
```

File: client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    public function setUp(): void
+    {
+        $this->caseRef = '1234567T';
+        $this->reportPdfUuid = '33ea0382-cfc9-4776-9036-667eeb68fa4b';
+        $this->expectedSupportingDocumentUuid = '9c0cb55e-718d-4ffb-9599-f3164e12dbdb';
+        $this->client = new GuzzleClient();
+        $this->signer = self::prophesize(RequestSigner::class);
+        $this->baseUrl = getenv('PACT_MOCK_SERVER_HOST');
+        $this->serializer = (self::bootKernel(['debug' => false]))->getContainer()->get('serializer');
+        $this->sut = new SiriusApiGatewayClient($this->client, $this->signer->reveal(), 'http://' . $this->baseUrl, $this->serializer);
```
Comment (gregtyler) [inline client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php]: Can simplify this a bit since there's no need to set them as properties (as they're never reused):
```suggestion
        $client = new GuzzleClient();
        $this->signer = self::prophesize(RequestSigner::class);
        $baseUrl = getenv('PACT_MOCK_SERVER_HOST');
        $serializer = (self::bootKernel(['debug' => false]))->getContainer()->get('serializer');
        $this->sut = new SiriusApiGatewayClient($client, $this->signer->reveal(), 'http://' . $baseUrl, $serializer);
```

File: client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                        'date_submitted' => $matcher->dateTimeWithMillisISO8601(),
+                        'type' => $matcher->regex('PF', 'PF|HW|NDR'),
+                        'submission_id' => $matcher->integer(9876)
+                    ]
+                ],
+            ]);
+
+        // Create a configuration that reflects the server that was started. You can create a custom MockServerConfigInterface if needed.
+        $config  = new MockServerEnvConfig();
+        $builder = new InteractionBuilder($config);
```
Comment (gregtyler) [inline client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php]: Might be worth moving this into the constructor as `$this->builder` to avoid duplication?

File: client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+        return $builder;
+    }
+
+    public function uploadProvider()
+    {
+        return [
+
+        ];
+    }
```
Comment (gregtyler) [inline client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php]: Looks like this isn't necessary?
```suggestion
```

File: client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+use DigidepsTests\Helpers\SiriusHelpers;
+use GuzzleHttp\Client;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Response;
+use Prophecy\Argument;
+use Prophecy\Prophecy\ObjectProphecy;
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+use Symfony\Component\Serializer\Serializer;
+
+class SiriusApiGatewayClientTest extends KernelTestCase
```
Comment (gregtyler) [inline client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php]: Is this file still needed now we have the other `SiriusApiGatewayClientTest`?

File: client/tests/phpunit/phpunit.xml [hunk 9]
```diff
@@ -10,7 +10,7 @@
          convertNoticesToExceptions="true"
          convertWarningsToExceptions="true"
         stopOnError="true"
-        stopOnFailure="true"
+        stopOnFailure="false"
```
Comment (gregtyler) [inline client/tests/phpunit/phpunit.xml]: Presumably this was just for testing?
```suggestion
        stopOnFailure="true"
```

File: client/src/AppBundle/Service/Client/Sirius/SiriusDocumentUpload.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
+<?php declare(strict_types=1);
+
+
+namespace AppBundle\Service\Client\Sirius;
```
Comment (gregtyler) [inline client/src/AppBundle/Service/Client/Sirius/SiriusDocumentUpload.php]: I think this should be in `AppBundle\Model\Sirius` since it's a model, not a service (same with the others)

File: client/tests/phpunit/Helpers/FileHelpers.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+     * @param string $fileLocation
+     * @param string $originalName
+     * @param string $mimeType
+     * @return UploadedFile
+     */
+    static public function generateUploadedFile(string $fileLocation, string $originalName, string $mimeType)
+    {
+        //@TODO drop the /../ file path when projectDir works as expected in Symfony 4+
+        $projectDir = (self::bootKernel(['debug' => false]))->getProjectDir();
+        $location = sprintf('%s/../%s', $projectDir, $fileLocation);
```
[the remark this answers is not in the document]
Comment (acsauk-mojd) [reply client/tests/phpunit/Helpers/FileHelpers.php]: I went down the route of trying to mock out the chain of calls down `UploadedFile` and its parent `File` but part of the `File` constructor checks that a real file exists at a path so I couldn't think of how to get around mocking that...open to suggestions if you've managed it in the past 🤷‍♂

File: client/src/AppBundle/Service/DocumentSyncService.php (hunk 3 above)
Comment (acsauk-mojd) [reply client/src/AppBundle/Service/DocumentSyncService.php]: I remember the day I was working on the error handling I was having an existential crisis about returning or throwing errors. I suspect I ended up on this solution as a temporary one and never came back to sorting it properly. I like your suggestion though so will go with that 👍

File: client/src/AppBundle/Entity/Report/Document.php (hunk 4 above)
Comment (acsauk-mojd) [reply client/src/AppBundle/Entity/Report/Document.php]: Ah I hadn't considered that - I've updated to use a filtered list now.

File: client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php (hunk 7 above)
Comment (acsauk-mojd) [reply client/tests/phpunit/Pact/SiriusApiGatewayClientTest.php]: Yup!

File: client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php (hunk 8 above)
Comment (acsauk-mojd) [reply client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php]: Hmmmm I guess we are doubling up a little here. I'll leave it with the get request test in for now. We'll need to update that to Pact too. I'll add a ticket for it.

File: client/tests/phpunit/phpunit.xml (hunk 9 above)
Comment (acsauk-mojd) [reply client/tests/phpunit/phpunit.xml]: Ah yeah, I like to see all the broken tests on a change rather than stopping at the first. I'll revert.

File: client/src/AppBundle/Service/Client/Sirius/SiriusDocumentUpload.php (hunk 10 above)
Comment (acsauk-mojd) [reply client/src/AppBundle/Service/Client/Sirius/SiriusDocumentUpload.php]: Agree! Moving over now.

File: client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php (hunk 8 above)
Comment (gregtyler) [reply client/tests/phpunit/Service/Client/Sirius/SiriusApiGatewayClientTest.php]: Fair enough 👍

File: client/tests/phpunit/Entity/Report/DocumentTest.php [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        $newReportPdfDocument->setReportSubmission($newReportPdfSubmission);
+        $supportingDocumentSubmittedWithReport->setReportSubmission($oldReportPdfSubmission);
+        $supportingDocumentSubmittedAfterReport->setReportSubmission($supportingDocumentSubmission);
+
+        $report->setReportSubmissions([$newReportPdfSubmission, $oldReportPdfSubmission, $supportingDocumentSubmission]);
+
+        self::assertEquals($oldReportPdfSubmission, $oldReportPdfDocument->getPreviousReportPdfSubmission());
+        self::assertEquals($oldReportPdfSubmission, $newReportPdfDocument->getPreviousReportPdfSubmission());
+        self::assertEquals($oldReportPdfSubmission, $supportingDocumentSubmittedWithReport->getPreviousReportPdfSubmission());
+        self::assertEquals($oldReportPdfSubmission, $supportingDocumentSubmittedAfterReport->getPreviousReportPdfSubmission());
```
Comment (gregtyler) [inline client/tests/phpunit/Entity/Report/DocumentTest.php]: Should this be `$newReportPdfSubmission`?

File: client/src/AppBundle/Service/DocumentSyncService.php (hunk 3 above)
Comment (gregtyler) [reply client/src/AppBundle/Service/DocumentSyncService.php]: Looks great 👍

File: client/src/AppBundle/Service/DocumentSyncService.php (hunk 2 above)
Comment (gregtyler) [reply client/src/AppBundle/Service/DocumentSyncService.php]: Needs updating a few other places in that file, sorry didn't clock that 🤦‍♂

File: client/tests/phpunit/Helpers/FileHelpers.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
+<?php declare(strict_types=1);
+
+
+namespace DigidepsTests\Helpers;
+
+
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+class FileHelpers extends KernelTestCase
```
Comment (gregtyler) [inline client/tests/phpunit/Helpers/FileHelpers.php]: Don't think this file is needed any more?

File: client/src/AppBundle/Service/DocumentSyncService.php (hunk 2 above)
Comment (acsauk-mojd) [reply client/src/AppBundle/Service/DocumentSyncService.php]: My bad! Updated now.

File: client/tests/phpunit/Entity/Report/DocumentTest.php (hunk 12 above)
Comment (acsauk-mojd) [reply client/tests/phpunit/Entity/Report/DocumentTest.php]: This was more to show that when a new submission is made we still get the previous submission back.

**Comment by acsauk-mojd** [comment]
Confirmed with integrations (https://mojdt.slack.com/archives/CBKSZGH0C/p1585070051112700) that they don't have the error handling code in place yet so I'm proposing we leave the Pact tests as they are with just the happy path.
